jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/TransformingFlusher.java

```diff
@@ -0,0 +1,168 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2019 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.internal;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.websocket.core.Frame;
+
+public abstract class TransformingFlusher
+{
+    private static final Logger LOG = Log.getLogger(TransformingFlusher.class);
+
+    private final Queue<FrameEntry> entries = new ArrayDeque<>();
+    private final IteratingCallback flusher = new Flusher();
+    private Throwable failure;
+
+    /**
+     * Used to iteratively transform a Frame.
+     * @param frame the frame to transform.
+     * @param callback used to signal to start processing again.
+     * @param batch whether this frame can be batched.
+     * @param first if this is the first time this particular frame has been given to transform.
+     * @return true when finished processing this frame.
+     */
+    protected abstract boolean transform(Frame frame, Callback callback, boolean batch, boolean first);
```

Comment:
How does the transformer keep track of where it is in the frame that it is transforming?  Does it alter the payload position?
If so, does it really need a boolean return?   Do normal sends alter the payload buffer of a frame they are sending?  If not then I'm not sure we should either... so perhaps the return should be an opaque state with a null state indicating completion.    Fragment state could just be an Integer and it would be passed back in as an object (or perhaps the class is templated over Integer?)

Reply:
The `TransformingFlusher` does not need to keep track of any state of the frame. Any extension of the `TransformingFlusher` can keep track of any state it needs to. Not sure why we need this fragment state you are talking about. The boolean return makes it more flexible, it does not need to know or make any assumptions what the implementation is doing with the frame.

With this implementation we could do something like this to repeat a frame multiple times. Never altering the payload of the original frame.
```java
public class RepeatingFlusher extends TransformingFlusher
{
    private final WebSocketCoreSession _session;
    private final int _numRepeats;
    private int _repeats = 0;

    public RepeatingFlusher(WebSocketCoreSession session, int numRepeats)
    {
        _session = session;
        _numRepeats = numRepeats;
    }

    @Override
    protected boolean transform(Frame frame, Callback callback, boolean batch, boolean first)
    {
        // Send this frame with sliced payload.
        Frame copy = Frame.copyWithoutPayload(frame);
        copy.setPayload(frame.getPayload().slice());
        _session.sendFrame(copy, callback, batch);

        if (_repeats == _numRepeats)
        {
            // Ready to receive next frame.
            _repeats = 0;
            return true;
        }

        // Send this frame again once callback is succeeded.
        _repeats++;
        return false;
    }
}
```

Reply:
How did it remember partiality consumed frames? If it is stateful, then why is the frame passed in?